### PR TITLE
Add Oam Designer

### PR DESCRIPTION
**Pull request by shinespeciall**
Add Oam Designer

### CODE REVIEW

File: Dialog/SpritesEditorDialog.cpp [hunk 1]
```diff
@@ -341,6 +360,107 @@ LevelComponents::EntitySet *SpritesEditorDialog::GetCurEntitySetPtr(bool createN
...
+/// Render OAM Set based on the ListView data
+/// </summary>
+/// <param name="selectrow">
+/// select a row after re-render oam graphic, leave it blank to set default value 0
+/// </param>
+void SpritesEditorDialog::RenderOamSet(int selectrow)
+{
+    int rowcount = ListViewItemModel->rowCount();
+    if (!rowcount) return;
+    unsigned short *nakedOAMData = new unsigned short[rowcount * 3];
```
Comment (Goldensunboy) [inline Dialog/SpritesEditorDialog.cpp:373]: Do not use new for this

File: Dialog/SpritesEditorDialog.cpp [hunk 2]
```diff
@@ -341,6 +360,107 @@ LevelComponents::EntitySet *SpritesEditorDialog::GetCurEntitySetPtr(bool createN
...
+}
+
+/// <summary>
+/// Helper function used to handle selection changed in ListView
+/// </summary>
+void SpritesEditorDialog::handleSelectionChanged()
+{
+    unsigned short nakedOAMData[3];
+    QStringList CurloadtableStrData = ListViewItemModel->item(SelectedRow_ListView)->text().split(QChar(' '), Qt::SkipEmptyParts);
+    nakedOAMData[0] = CurloadtableStrData.at(0).toUInt(nullptr, 16);
```
Comment (Goldensunboy) [inline Dialog/SpritesEditorDialog.cpp:447]: Use a for loop

File: Dialog/SpritesEditorDialog.cpp [hunk 3]
```diff
@@ -341,6 +360,107 @@ LevelComponents::EntitySet *SpritesEditorDialog::GetCurEntitySetPtr(bool createN
...
+/// <return>
+/// single OAM data QString
+/// </return>
+QString SpritesEditorDialog::GenerateOAMString()
+{
+    unsigned short tmpOAMData[3] = {0, 0, 0};
+    int shapesizedata = ui->comboBox_OamShapeType->currentIndex();
+    int yvalue = (ui->spinBox_OamY->value() >= 0) ? ui->spinBox_OamY->value() : 0x100 + ui->spinBox_OamY->value();
+    tmpOAMData[0] = yvalue |
+            (shapesizedata & 0xC) >> 2 << 14 |
```
Comment (Goldensunboy) [inline Dialog/SpritesEditorDialog.cpp:408]: Can you add an extra set of parenthesis here to make the order of operations more clear between >> and <<

File: Dialog/SpritesEditorDialog.cpp [hunk 4]
```diff
@@ -559,3 +679,116 @@ void SpritesEditorDialog::on_pushButton_SwapPal_clicked()
     SetSelectedSpriteTile(0);
     RenderSpritesetTileMapAndResetLoadTable();
 }
+
+/// <summary>
+/// Reset Oam designer with new OAM data string when clicking pushButton_ResetAllOamData
+/// </summary>
+void SpritesEditorDialog::on_pushButton_ResetAllOamData_clicked()
+{
+    QStringList List_strs = QInputDialog::getText(this, tr("InputBox"),
```
Comment (Goldensunboy) [inline Dialog/SpritesEditorDialog.cpp:688]: let's use camelCase for the variable name

File: Dialog/SpritesEditorDialog.cpp [hunk 5]
```diff
@@ -559,3 +679,116 @@ void SpritesEditorDialog::on_pushButton_SwapPal_clicked()
...
+    if (List_strs.size() % 3)
+    {
+        QMessageBox::critical(this, tr("Error"), QString(tr("Data number should be multiple of 3 !")));
+        return;
+    }
+    if (ListViewItemModel)
+    {
+        ListViewItemModel->clear();
+        delete ListViewItemModel;
+        ListViewItemModel = nullptr;
```
Comment (Goldensunboy) [inline Dialog/SpritesEditorDialog.cpp:702]: Is this line necessary?

File: LevelComponents/Entity.cpp [hunk 6]
```diff
@@ -255,6 +255,100 @@ namespace LevelComponents
...
+    /// </summary>
+    /// <param name="oamNum">
+    /// the number of oam.
+    /// </param>
+    /// <param name="nakedOAMarray">
+    /// Naked OAM array without oam number as the first element
+    /// </param>
+    /// <param name="noReferrneceBox">
+    /// Don't draw the yellow referrence Box if set this to true
+    /// </param>
```
Comment (Goldensunboy) [inline LevelComponents/Entity.cpp:269]: missing `<returns>`

File: LevelComponents/Entity.cpp [hunk 7]
```diff
@@ -255,6 +255,100 @@ namespace LevelComponents
...
+    /// <summary>
+    /// Render oam array in test mode.
+    /// </summary>
+    /// <param name="oamNum">
+    /// the number of oam.
+    /// </param>
+    /// <param name="nakedOAMarray">
+    /// Naked OAM array without oam number as the first element
+    /// </param>
+    /// <param name="noReferrneceBox">
```
Comment (Goldensunboy) [inline LevelComponents/Entity.cpp:267]: noReferenceBox

File: LevelComponents/Entity.cpp [hunk 8]
```diff
@@ -255,6 +255,100 @@ namespace LevelComponents
...
+    /// Don't draw the yellow referrence Box if set this to true
+    /// </param>
+    QImage Entity::TestRenderOams(int oamNum, unsigned short *nakedOAMarray, bool noReferrneceBox)
```
Comment (Goldensunboy) [inline LevelComponents/Entity.cpp:270]: The function has test in the name. Is it used for testing? Or something else?

If it's used for testing, we should not put it in this file. If it's not used for testing, we should change the name.

File: LevelComponents/Entity.cpp [hunk 9]
```diff
@@ -255,6 +255,100 @@ namespace LevelComponents
...
+            unsigned short attr0 = nakedOAMarray[3 * i];
+            unsigned short attr1 = nakedOAMarray[3 * i + 1];
+            unsigned short attr2 = nakedOAMarray[3 * i + 2];
+            // skip priority reset and Char Id (tiles ID) reset
+            // skip orientation change
+            // skip affine transformation (double size)
+            // skip alpha blend
+
+            // Obtain the tile parameters for the OAM tile
+            struct OAMTile *newOAM = new struct OAMTile();
```
Comment (Goldensunboy) [inline LevelComponents/Entity.cpp:287]: This is a memory leak. Change the vector to not use pointers

File: LevelComponents/Entity.cpp [hunk 10]
```diff
@@ -255,6 +255,100 @@ namespace LevelComponents
...
+            newOAM->Xoff = (attr1 & 0xFF) - (attr1 & 0x100); // Offset of OAM tile from entity origin
+            newOAM->Yoff = (attr0 & 0x7F) - (attr0 & 0x80); // they are signed char
+            newOAM->xFlip = (attr1 & (1 << 0xC)) != 0;
```
Comment (Goldensunboy) [inline LevelComponents/Entity.cpp:290]: You can remove `!= 0`

File: LevelComponents/Entity.cpp [hunk 11]
```diff
@@ -255,6 +255,100 @@ namespace LevelComponents
...
+                    entityTile->deltaY = y * 8;
+                    int offsetID = tileID + y * 0x20 + x;
+                    if (offsetID >= 0x200) offsetID = offsetID - 0x200; // the game engine use char Id with offset in the oam data
+                    if (offsetID >= tile8x8data.size()) continue; // this data have problems, just skip it
+                    if (tile8x8data[offsetID] == blankTile) continue; // skip dummy tiles if used
+                    Tile8x8 *newTile = new Tile8x8(tile8x8data[offsetID]);
+                    if (palNum > 7) palNum = palNum - 8; // the game engine use pal Id with offset in the oam data
+                    newTile->SetPaletteIndex(palNum);
+                    entityTile->objTile = newTile;
+                    newOAM->tile8x8.push_back(entityTile);
```
Comment (Goldensunboy) [inline LevelComponents/Entity.cpp:315]: Where does entityTile get used later? This also seems like a memory leak and does not need to be a pointer

File: LevelComponents/Entity.cpp [hunk 12]
```diff
@@ -255,6 +255,100 @@ namespace LevelComponents
...
+            // x + 8, y + 16, this works in the room rendering to match the box
+            p.drawImage(ot->Xoff + 512 + 8, ot->Yoff + 256 + 16, ot->Render());
+            delete ot;
+        }
+        tmpOAMTiles.clear();
+
+        // Draw position referrence Box
+        if (!noReferrneceBox)
+        {
+            QPen EntityBoxPen = QPen(QBrush(QColor(0xFF, 0xFF, 0, 0xFF)), 2);
```
Comment (Goldensunboy) [inline LevelComponents/Entity.cpp:343]: Let's `#define` the color (and construct the color from an RGBA string instead of components)

**Review by Goldensunboy — CHANGES_REQUESTED** [review]

File: Dialog/SpritesEditorDialog.cpp (hunk 1 above)
Comment (shinespeciall) [reply Dialog/SpritesEditorDialog.cpp:373]: no, its size will change every time

Comment (Goldensunboy) [reply Dialog/SpritesEditorDialog.cpp:373]: That is fine. You should only use dynamic memory if the data should persist after this function. If it is used only in this function, then do not use `new`

Comment (Goldensunboy) [reply Dialog/SpritesEditorDialog.cpp:373]: So, please change this to `unsigned short nakedOAMData[rowcount * 3];`

File: Dialog/SpritesEditorDialog.cpp (hunk 2 above)
Comment (shinespeciall) [reply Dialog/SpritesEditorDialog.cpp:447]: done.

File: Dialog/SpritesEditorDialog.cpp (hunk 1 above)
Comment (shinespeciall) [reply Dialog/SpritesEditorDialog.cpp:373]: done with QVector and for loop

File: Dialog/SpritesEditorDialog.cpp (hunk 3 above)
Comment (shinespeciall) [reply Dialog/SpritesEditorDialog.cpp:408]: done.

File: Dialog/SpritesEditorDialog.cpp (hunk 4 above)
Comment (shinespeciall) [reply Dialog/SpritesEditorDialog.cpp:688]: done

File: Dialog/SpritesEditorDialog.cpp (hunk 5 above)
Comment (shinespeciall) [reply Dialog/SpritesEditorDialog.cpp:702]: deleted.

File: LevelComponents/Entity.cpp (hunk 6 above)
Comment (shinespeciall) [reply LevelComponents/Entity.cpp:269]: added.

File: LevelComponents/Entity.cpp (hunk 7 above)
Comment (shinespeciall) [reply LevelComponents/Entity.cpp:267]: done.

File: LevelComponents/Entity.cpp (hunk 8 above)
Comment (shinespeciall) [reply LevelComponents/Entity.cpp:270]: replaced with RenderOAMPreview

File: LevelComponents/Entity.cpp (hunk 10 above)
Comment (shinespeciall) [reply LevelComponents/Entity.cpp:290]: done.

File: LevelComponents/Entity.cpp (hunk 9 above)
Comment (shinespeciall) [reply LevelComponents/Entity.cpp:287]: it is deleted in the render for loop

File: LevelComponents/Entity.cpp (hunk 11 above)
Comment (shinespeciall) [reply LevelComponents/Entity.cpp:315]: it is deleted in the deconstructor of OAMTile which can be deleted when deleting OAMTile, which is done in the render for loop

File: LevelComponents/Entity.cpp (hunk 12 above)
Comment (shinespeciall) [reply LevelComponents/Entity.cpp:343]: done.

File: LevelComponents/Entity.cpp [hunk 13]
```diff
@@ -338,9 +343,9 @@ namespace LevelComponents
         tmpOAMTiles.clear();
 
         // Draw position referrence Box
-        if (!noReferrneceBox)
+        if (!noReferenceBox)
         {
-            QPen EntityBoxPen = QPen(QBrush(QColor(0xFF, 0xFF, 0, 0xFF)), 2);
+            QPen EntityBoxPen = QPen(QBrush(QColor(entityReferBoxRGB)), 2);
```
Comment (Goldensunboy) [inline LevelComponents/Entity.cpp:348]: change to EntityReferenceBoxRGB

Comment (shinespeciall) [reply LevelComponents/Entity.cpp]: done

**Review by Goldensunboy — APPROVED** [review]

**Pull request merged**